### PR TITLE
[Bugfix][Model Runner V2] Preserve sampling masks in batch-sharded sampling

### PR DESCRIPTION
**Pull request by waizuichougou**
## Purpose

This is a regression introduced by the merge of #50465. The earlier #49577
change added sampling-distribution replay and `--return-sampling-mask`, while
#50465 added batch-sharded sampler-output gathering without forwarding the
existing sampling-mask metadata through that path.

When Model Runner V2 uses batch-sharded sampling with tensor parallelism, the
sampler can produce `SamplingMaskTensors` when `--return-sampling-mask` is
enabled, but `gather_sampler_output()` drops them while gathering the sampled
tokens and logprobs. The request still completes successfully, but the
`sampling_mask` field in the generate response is `null`.

This makes sampling-distribution replay unusable for this explicitly supported
combination, which affects RL and other consumers that need the post-processing
token support. The default sampling path and batch-sharded sampling without
sampling masks are unchanged.

The fix pads each rank's packed masks and counts to the common per-rank request
capacity, all-gathers both tensors only when sampling masks are requested, and
uses the existing owner-to-global request mapping to restore the original batch
order. A regression test covers both a rank that owns no requests and mixed-owner
request ordering.

## Duplicate-work check

Duplicate-work check: searched open PRs and issues for `batch-sharded sampling
mask`, `return_sampling_mask`, and related terms on 2026-08-26; no existing fix
was found. This PR addresses the integration gap between #49577 and #50465:
the sampling-mask output path exists, and batch-sharded sampling exists, but
their sampler-output gathering path did not preserve the mask metadata.

### Reproduction

Start a server with Model Runner V2, tensor parallelism, batch-sharded sampling,
and sampling-mask output enabled:

```bash
MODEL_ID=Qwen/Qwen3.5-0.8B-Base
VLLM_USE_V2_MODEL_RUNNER=1 \
  vllm serve "$MODEL_ID" \
  --host 127.0.0.1 \
  --port 18080 \
  --tensor-parallel-size 2 \
  --enable-batch-sharded-sampling \
  --return-sampling-mask \
  --logprobs-mode processed_logprobs \
  --max-model-len 256 \
  --max-num-seqs 2 \
  --gpu-memory-utilization 0.35 \
  --enforce-eager \
  --no-enable-prefix-caching \
  --trust-remote-code
```

Send a non-greedy request with `top_k > 0` through the token-in/token-out
endpoint:

```bash
curl -sS http://127.0.0.1:18080/inference/v1/generate \
  -H 'Content-Type: application/json' \
  -d '{
    "token_ids": [151644, 872, 198],
    "sampling_params": {
      "temperature": 0.8,
      "top_k": 5,
      "top_p": 0.9,
      "max_tokens": 2,
      "seed": 123
    },
    "stream": false
  }'
```

The observed unpatched baseline returned HTTP 200 and completed generation, but
the choice contained:

```json
"sampling_mask": null
```

The fixed run returned HTTP 200 with the following output fields:

```json
"token_ids": [410, 149852],
"sampling_mask": [
  [28, 410, 590, 8442],
  [220, 271, 149852, 149910]
]
```

For this request, the internal CSR representation was:

```text
prompt token_ids:       [151644, 872, 198]
generated token_ids:    [410, 149852]
mask token_ids (flat):  [28, 410, 590, 8442, 220, 271, 149852, 149910]
offsets:                [0, 4, 8]
counts:                 [4, 4]
```

Both generated tokens are members of their corresponding support set, and each
support set has four tokens, which is at most `top_k=5`. The generated token IDs
in the baseline and fixed runs are not used as a quality comparison; the
metadata-preservation check is the presence, alignment, and cardinality of the
support sets.

## Test Plan

- Add a regression test for mask/count gathering, request-order restoration, an
  empty local shard, and mixed-owner ordering.
- Run the focused batch-sharded sampling test file.
- Run the repository pre-commit checks for all changed files.
- Run the exact `vllm serve` reproduction above and send the seeded request to
  `/inference/v1/generate`.

## Test Result

- `pytest -q tests/v1/worker/test_gpu_batch_shard.py`: 18 passed.
- `ruff check` and `ruff format --check`: passed.
- Repository pre-commit checks for the changed files: passed, including ruff,
  mypy, SPDX, forbidden-import, and configuration checks.
- End-to-end `vllm serve` verification: the unpatched baseline returned HTTP
  200 with `sampling_mask: null`; the fixed seeded run (`seed=123`) returned
  HTTP 200 with `token_ids=[410, 149852]`, `offsets=[0, 4, 8]`,
  `counts=[4, 4]`, and two four-token support sets. The generated tokens were
  present in their matching support sets.

No documentation update is required because this change restores the behavior
of the existing `--return-sampling-mask` option.

## Model Evaluation

Not applicable. This change does not alter sampled token IDs, logits, logprobs,
or filtering behavior; it only preserves already-produced sampling-mask
metadata through batch-sharded result gathering. Traditional quality or
perplexity evaluation would not detect this metadata-only regression. The
baseline/fixed serving comparison above validates the affected output contract.

## AI Assistance Disclosure

This change was developed with AI assistance.

### CODE REVIEW

**Comment by waizuichougou** [comment]
With #54901 now merged, I wanted to briefly follow up on this PR. Its description notes that this fix could lift the current restriction on combining --return-sampling-mask with --enable-batch-sharded-sampling. This seems to indicate that the two changes are complementary, and that this PR could help complete support for this sampling-mask configuration.
